# Homework Assignment - Lesson 7: String Manipulation and Date/Time Data

**Student Name:** [Vi McAllister]

**Student ID:** [fks355]

**Date Submitted:** [10/13]

**Due Date:** [10/12]

---

## Objective

Master string manipulation with `stringr` and date/time operations with `lubridate` for real-world business data cleaning and analysis.

## Learning Goals

By completing this assignment, you will:
- Clean and standardize messy text data using `stringr` functions
- Parse and manipulate dates using `lubridate` functions
- Extract information from text and dates for business insights
- Combine string and date operations for customer segmentation
- Create business-ready reports from raw data

## Instructions

- Complete all tasks in this notebook
- Write your code in the designated TODO sections
- Use the pipe operator (`%>%`) wherever possible
- Add comments explaining your logic
- Run all cells to verify your code works
- Answer all reflection questions

## Datasets

You will work with three CSV files:
- `customer_feedback.csv` - Customer reviews with messy text
- `transaction_log.csv` - Transaction records with dates
- `product_catalog.csv` - Product descriptions needing standardization

---

## Part 1: Data Import and Initial Exploration

**Business Context:** Before cleaning data, you must understand its structure and quality issues.

**Your Tasks:**
1. Load required packages (`tidyverse` and `lubridate`)
2. Import all three CSV files from the `data/` directory
3. Examine the structure and identify data quality issues
4. Display sample rows to understand the data

In [2]:
# Task 1.1: Load Required Packages
# TODO: Load tidyverse (includes stringr)
library(tidyverse)

# TODO: Load lubridate
library(lubridate)


cat("✅ Packages loaded successfully!\n")

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.1     ✔ stringr   1.5.2
✔ ggplot2   4.0.0     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.1.0     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


✅ Packages loaded successfully!


In [3]:
# Task 1.2: Import Datasets
getwd()
setwd('/workspaces/assignment-1-version-3-ViMcAllister/data')
# TODO: Import customer_feedback.csv into a variable called 'feedback'
feedback <- read.csv('/workspaces/assignment-1-version-3-ViMcAllister/data/customer_feedback (1).csv')

# TODO: Import transaction_log.csv into a variable called 'transactions'
transactions <- read.csv('/workspaces/assignment-1-version-3-ViMcAllister/data/transaction_log.csv')

# TODO: Import product_catalog.csv into a variable called 'products'
products <- read.csv('/workspaces/assignment-1-version-3-ViMcAllister/data/product_catalog.csv')

cat("✅ Data imported successfully!\n")
cat("Feedback rows:", nrow(feedback), "\n")
cat("Transaction rows:", nrow(transactions), "\n")
cat("Product rows:", nrow(products), "\n")

[1] "/workspaces/assignment-1-version-3-ViMcAllister/assignment/Homework"

✅ Data imported successfully!
Feedback rows: 100 
Transaction rows: 150 
Product rows: 75 


In [4]:
# Task 1.3: Initial Data Exploration

cat("=== CUSTOMER FEEDBACK DATA ===\n")
# TODO: Display structure of feedback using str()
print(str(feedback))


# TODO: Display first 5 rows of feedback
head(feedback,5)

cat("\n=== TRANSACTION DATA ===\n")
# TODO: Display structure of transactions
print(str(transactions))

# TODO: Display first 5 rows of transactions
head(transactions,5)

cat("\n=== PRODUCT CATALOG DATA ===\n")
# TODO: Display structure of products
print(str(products))

# TODO: Display first 5 rows of products
head(products,5)


=== CUSTOMER FEEDBACK DATA ===
'data.frame':	100 obs. of  5 variables:
 $ FeedbackID   : int  1 2 3 4 5 6 7 8 9 10 ...
 $ CustomerID   : int  12 40 34 1 47 13 13 37 49 23 ...
 $ Feedback_Text: chr  "Highly recommend this item" "   Excellent service   " "Poor quality control" "average product, nothing special" ...
 $ Contact_Info : chr  "bob.wilson@test.org" "555-123-4567" "jane_smith@company.com" "jane_smith@company.com" ...
 $ Feedback_Date: chr  "2024-02-23" "2024-01-21" "2023-09-02" "2023-08-21" ...
NULL


FeedbackID CustomerID Feedback_Text                    Contact_Info          
1 1          12         Highly recommend this item       bob.wilson@test.org   
2 2          40            Excellent service             555-123-4567          
3 3          34         Poor quality control             jane_smith@company.com
4 4           1         average product, nothing special jane_smith@company.com
5 5          47         AMAZING customer support!!!      555-123-4567          
  Feedback_Date
1 2024-02-23   
2 2024-01-21   
3 2023-09-02   
4 2023-08-21   
5 2023-04-24


=== TRANSACTION DATA ===
'data.frame':	150 obs. of  5 variables:
 $ LogID               : int  1 2 3 4 5 6 7 8 9 10 ...
 $ CustomerID          : int  26 21 12 6 32 27 31 30 31 13 ...
 $ Transaction_DateTime: chr  "4/5/24 14:30" "3/15/24 14:30" "3/15/24 14:30" "3/20/24 9:15" ...
 $ Amount              : num  277 175 252 215 269 ...
 $ Status              : chr  "Pending" "Pending" "Pending" "Pending" ...
NULL


LogID CustomerID Transaction_DateTime Amount Status   
1 1     26         4/5/24 14:30         277.22 Pending  
2 2     21         3/15/24 14:30        175.16 Pending  
3 3     12         3/15/24 14:30        251.71 Pending  
4 4      6         3/20/24 9:15         214.98 Pending  
5 5     32         3/20/24 9:15         268.91 Completed


=== PRODUCT CATALOG DATA ===
'data.frame':	75 obs. of  5 variables:
 $ ProductID          : int  1 2 3 4 5 6 7 8 9 10 ...
 $ Product_Description: chr  "Apple iPhone 14 Pro - 128GB - Space Black" "samsung galaxy s23 ultra 256gb" "Apple iPhone 14 Pro - 128GB - Space Black" "Apple iPhone 14 Pro - 128GB - Space Black" ...
 $ Category           : chr  "TV" "TV" "Audio" "Shoes" ...
 $ Price              : num  964 1817 853 649 586 ...
 $ In_Stock           : chr  "Limited" "Yes" "Yes" "Yes" ...
NULL


ProductID Product_Description                       Category    Price  
1 1         Apple iPhone 14 Pro - 128GB - Space Black TV           963.53
2 2         samsung galaxy s23 ultra 256gb            TV          1817.44
3 3         Apple iPhone 14 Pro - 128GB - Space Black Audio        852.79
4 4         Apple iPhone 14 Pro - 128GB - Space Black Shoes        648.58
5 5         samsung galaxy s23 ultra 256gb            Electronics  586.35
  In_Stock
1 Limited 
2 Yes     
3 Yes     
4 Yes     
5 Limited

## Part 2: String Cleaning and Standardization

**Business Context:** Product names and feedback text often have inconsistent formatting that prevents accurate analysis.

**Your Tasks:**
1. Clean product names (remove extra spaces, standardize case)
2. Standardize product categories
3. Clean customer feedback text
4. Extract customer names from feedback

**Key Functions:** `str_trim()`, `str_squish()`, `str_to_lower()`, `str_to_upper()`, `str_to_title()`

In [5]:
# Task 2.1: Clean Product Names
# TODO: Create a new column 'product_name_clean' that:
#   - Removes leading/trailing whitespace using str_trim()
#   - Converts to Title Case using str_to_title()

products_clean <- products  %>% 
  mutate(
  product_name = str_trim(ProductID),
  product_name_clean = str_to_title(ProductID)

  )


# Display before and after
cat("Product Name Cleaning Results:\n")
products_clean %>%
  select(product_name, product_name_clean) %>%
  head(10) %>%
  print()

Product Name Cleaning Results:
   product_name product_name_clean
1             1                  1
2             2                  2
3             3                  3
4             4                  4
5             5                  5
6             6                  6
7             7                  7
8             8                  8
9             9                  9
10           10                 10


In [6]:
# Task 2.2: Standardize Product Categories
# TODO: Create a new column 'category_clean' that:
#   - Converts category to Title Case
#   - Removes any extra whitespace

products_clean <- products_clean %>%
  mutate(
  category_clean = str_trim(str_to_title(Category))
    
  )

# Show unique categories before and after
cat("Original categories:\n")
print(unique(products$category))

cat("\nCleaned categories:\n")
print(unique(products_clean$category_clean))

Original categories:
NULL

Cleaned categories:
[1] "Tv"          "Audio"       "Shoes"       "Electronics" "Computers"  


In [7]:
# Task 2.3: Clean Customer Feedback Text
# TODO: Create a new column 'feedback_clean' that:
#   - Converts text to lowercase using str_to_lower()
#   - Removes extra whitespace using str_squish()

feedback_clean <- feedback %>%
  mutate(
    feedback_text = str_to_lower(Feedback_Text),
  feedback_clean = str_squish(Feedback_Text)
  )

# Display sample
cat("Feedback Cleaning Sample:\n")
feedback_clean %>%
  select(feedback_text, feedback_clean) %>%
  head(5) %>%
  print()

Feedback Cleaning Sample:
                     feedback_text                   feedback_clean
1       highly recommend this item       Highly recommend this item
2             excellent service                   Excellent service
3             poor quality control             Poor quality control
4 average product, nothing special average product, nothing special
5      amazing customer support!!!      AMAZING customer support!!!


## Part 3: Pattern Detection and Extraction

**Business Context:** Identifying products with specific features and extracting specifications helps with inventory management and marketing.

**Your Tasks:**
1. Identify products with specific keywords (wireless, premium, gaming)
2. Extract numerical specifications from product names
3. Detect sentiment words in customer feedback
4. Extract email addresses from feedback

**Key Functions:** `str_detect()`, `str_extract()`, `str_count()`

In [8]:
# Task 3.1: Detect Product Features
# TODO: Create three new columns:
#   - is_wireless: TRUE if product name contains "wireless" (case-insensitive)
#   - is_premium: TRUE if product name contains "pro", "premium", or "deluxe"
#   - is_gaming: TRUE if product name contains "gaming" or "gamer"
# Hint: Use str_detect() with str_to_lower() for case-insensitive matching
# Hint: Use | (pipe) in regex for OR conditions

products_clean <- products_clean %>%
  mutate(
  is_wireless = str_detect(str_to_lower(Product_Description), "wireless"),
  is_premium = str_detect(str_to_lower(Product_Description), "pro|premium|deluxe"),
  is_gaming = str_detect(str_to_lower(Product_Description), "gaming|gamer")
  )

# Display results
cat("Product Feature Detection:\n")
products_clean %>%
  select(product_name_clean, is_wireless, is_premium, is_gaming) %>%
  head(10) %>%
  print()

# Summary statistics
cat("\nFeature Summary:\n")
cat("Wireless products:", sum(products_clean$is_wireless), "\n")
cat("Premium products:", sum(products_clean$is_premium), "\n")
cat("Gaming products:", sum(products_clean$is_gaming), "\n")

Product Feature Detection:
   product_name_clean is_wireless is_premium is_gaming
1                   1       FALSE       TRUE     FALSE
2                   2       FALSE      FALSE     FALSE
3                   3       FALSE       TRUE     FALSE
4                   4       FALSE       TRUE     FALSE
5                   5       FALSE      FALSE     FALSE
6                   6       FALSE       TRUE     FALSE
7                   7       FALSE      FALSE     FALSE
8                   8        TRUE      FALSE     FALSE
9                   9       FALSE      FALSE     FALSE
10                 10       FALSE      FALSE     FALSE

Feature Summary:
Wireless products: 17 
Premium products: 13 
Gaming products: 0 


In [9]:
# Task 3.2: Extract Product Specifications
# TODO: Create a new column 'size_number' that extracts the first number from product_name
# Hint: Use str_extract() with pattern "\\\\d+" to match one or more digits

products_clean <- products_clean %>%
  mutate(
  size_number = str_extract(product_name,"\\\\d+")
  )

# Display products with extracted sizes
cat("Extracted Product Specifications:\n")
products_clean %>%
  filter(!is.na(size_number)) %>%
  select(product_name_clean, size_number) %>%
  head(10) %>%
  print()

Extracted Product Specifications:
[1] product_name_clean size_number       
<0 rows> (or 0-length row.names)


In [10]:
# Task 3.3: Simple Sentiment Analysis
# TODO: Create three new columns:
#   - positive_words: count of positive words ("great", "excellent", "love", "amazing")
#   - negative_words: count of negative words ("bad", "terrible", "hate", "awful")
#   - sentiment_score: positive_words - negative_words
# Hint: Use str_count() to count pattern occurrences

feedback_clean <- feedback_clean %>%
  mutate(
  positive_words = str_count(Feedback_Text, "great|excellent|love|amazing"),
  negative_words = str_count(Feedback_Text, "bad|terrible|hate|awful"),
  sentiment_score = positive_words - negative_words
  )

# Display sentiment analysis results
cat("Sentiment Analysis Results:\n")
feedback_clean %>%
  select(feedback_clean, positive_words, negative_words, sentiment_score) %>%
  head(10) %>%
  print()

# Summary
cat("\nOverall Sentiment Summary:\n")
cat("Average sentiment score:", mean(feedback_clean$sentiment_score), "\n")
cat("Positive reviews:", sum(feedback_clean$sentiment_score > 0), "\n")
cat("Negative reviews:", sum(feedback_clean$sentiment_score < 0), "\n")

Sentiment Analysis Results:
                     feedback_clean positive_words negative_words
1        Highly recommend this item              0              0
2                 Excellent service              0              0
3              Poor quality control              0              0
4  average product, nothing special              0              0
5       AMAZING customer support!!!              0              0
6       AMAZING customer support!!!              0              0
7  average product, nothing special              0              0
8              good VALUE for money              0              0
9        Highly recommend this item              0              0
10       Highly recommend this item              0              0
   sentiment_score
1                0
2                0
3                0
4                0
5                0
6                0
7                0
8                0
9                0
10               0

Overall Sentiment Summary:
Average s

## Part 4: Date Parsing and Component Extraction

**Business Context:** Transaction dates need to be parsed and analyzed to understand customer behavior patterns.

**Your Tasks:**
1. Parse transaction dates from text to Date objects
2. Extract date components (year, month, day, weekday)
3. Identify weekend vs weekday transactions
4. Extract quarter and month names

**Key Functions:** `ymd()`, `mdy()`, `dmy()`, `year()`, `month()`, `day()`, `wday()`, `quarter()`

In [11]:
# Task 4.1: Parse Transaction Dates
# TODO: Create a new column 'date_parsed' that parses the transaction_date column
# Hint: Check the format of transaction_date first, then use ymd(), mdy(), or dmy()

transactions_clean <- transactions %>%
  mutate(
  transaction_date = Transaction_DateTime,
  date_parsed = ymd(transaction_date)
  )

# Verify parsing worked
cat("Date Parsing Results:\n")
transactions_clean %>%
  select(transaction_date, date_parsed) %>%
  head(10) %>%
  print()

Warning message:
“There was 1 warning in `mutate()`.
ℹ In argument: `date_parsed = ymd(transaction_date)`.
Caused by warning:
! All formats failed to parse. No formats found.”


Date Parsing Results:
      transaction_date date_parsed
1         4/5/24 14:30        <NA>
2        3/15/24 14:30        <NA>
3        3/15/24 14:30        <NA>
4         3/20/24 9:15        <NA>
5         3/20/24 9:15        <NA>
6         3/20/24 9:15        <NA>
7         3/20/24 9:15        <NA>
8        3/15/24 14:30        <NA>
9  25-03-2024 16:45:30        <NA>
10        4/5/24 14:30        <NA>


In [13]:
# Task 4.2: Extract Date Components
# TODO: Create the following new columns:
#   - trans_year: Extract year from date_parsed
#   - trans_month: Extract month number from date_parsed
#   - trans_month_name: Extract month name (use label=TRUE, abbr=FALSE)
#   - trans_day: Extract day of month from date_parsed
#   - trans_weekday: Extract weekday name (use label=TRUE, abbr=FALSE)
#   - trans_quarter: Extract quarter from date_parsed

transactions_clean <- transactions_clean %>%
  mutate(
trans_year = year(date_parsed),
trans_month = month(date_parsed),
trans_month_name = extract(date_parsed, label = TRUE, abbr = FALSE),
trans_day = day(date_parsed),
trans_weekday = wday(date_parsed, label = TRUE, abbr = FALSE)
  )

# Display results
cat("Date Component Extraction:\n")
transactions_clean %>%
  select(date_parsed, trans_month_name, trans_weekday, trans_quarter) %>%
  head(10) %>%
  print()

ERROR: Error in `mutate()`:
ℹ In argument: `trans_month_name = extract(date_parsed, label = TRUE,
  abbr = FALSE)`.
Caused by error in `UseMethod()`:
! no applicable method for 'extract' applied to an object of class "Date"


ERROR: Error in `mutate()`:
ℹ In argument: `trans_month_name = extract(date_parsed, label = TRUE,
  abbr = FALSE)`.
Caused by error in `extract()`:
! Arguments in `...` must be used.
✖ Problematic arguments:
• label = TRUE
• abbr = FALSE
ℹ Did you misspell an argument name?


In [14]:
# Task 4.3: Identify Weekend Transactions
# TODO: Create a new column 'is_weekend' that is TRUE if the transaction was on Saturday or Sunday
# Hint: Use wday() which returns 1 for Sunday and 7 for Saturday
# Hint: Use %in% c(1, 7) to check if day is weekend

transactions_clean <- transactions_clean %>%
  mutate(
  is_weekend = wday(date_parsed, label = 1, abbr = 7)
  )

# Summary
cat("Weekend vs Weekday Transactions:\n")
table(transactions_clean$is_weekend) %>% print()

cat("\nPercentage of weekend transactions:",
    round(sum(transactions_clean$is_weekend) / nrow(transactions_clean) * 100, 1), "%\n")

Weekend vs Weekday Transactions:

Sun Mon Tue Wed Thu Fri Sat 
  0   0   0   0   0   0   0 


ERROR: Error in Summary.ordered(structure(c(NA_integer_, NA_integer_, NA_integer_, : 'sum' not defined for ordered factors


## Part 5: Date Calculations and Customer Recency Analysis

**Business Context:** Understanding how recently customers transacted helps identify at-risk customers for re-engagement campaigns.

**Your Tasks:**
1. Calculate days since each transaction
2. Categorize customers by recency (Recent, Moderate, Old)
3. Identify customers who haven't transacted in 90+ days
4. Calculate average days between transactions per customer

**Key Functions:** `today()`, date arithmetic, `case_when()`

In [16]:
# Task 5.1: Calculate Days Since Transaction
# TODO: Create a new column 'days_since' that calculates days from date_parsed to today()
# Hint: Use as.numeric(today() - date_parsed)

transactions_clean <- transactions_clean %>%
  mutate(
days_since = as.numeric(today() - date_parsed)
  )

# Display results
cat("Days Since Transaction:\n")
transactions_clean %>%
  select(customer_name, date_parsed, days_since) %>%
  arrange(desc(days_since)) %>%
  head(10) %>%
  print()

Days Since Transaction:


ERROR: Error in `select()`:
! Can't select columns that don't exist.
✖ Column `customer_name` doesn't exist.


In [ ]:
# Task 5.2: Categorize by Recency
# TODO: Create a new column 'recency_category' using case_when():
#   - "Recent" if days_since <= 30
#   - "Moderate" if days_since <= 90
#   - "At Risk" if days_since > 90

transactions_clean <- transactions_clean %>%
  mutate(
recency_category = case_when(
  Recent <= 30 ~ 
  Moderate <= 90 ~ 
)
  )

# Display distribution
cat("Recency Category Distribution:\n")
table(transactions_clean$recency_category) %>% print()

# Show at-risk customers
cat("\nAt-Risk Customers (>90 days):\n")
transactions_clean %>%
  filter(recency_category == "At Risk") %>%
  select(customer_name, date_parsed, days_since) %>%
  arrange(desc(days_since)) %>%
  print()

## Part 6: Combined String and Date Operations

**Business Context:** Create personalized customer outreach messages based on purchase recency.

**Your Tasks:**
1. Extract first names from customer names
2. Create personalized messages based on recency
3. Analyze transaction patterns by weekday
4. Identify best customers (recent + high value)

**Key Functions:** Combine `str_extract()`, date calculations, `case_when()`, `group_by()`, `summarize()`

In [ ]:
# Task 6.1: Extract First Names and Create Personalized Messages
# TODO: Create two new columns:
#   - first_name: Extract first name from customer_name (everything before first space)
#   - personalized_message: Create message based on recency_category
#     * Recent: "Hi [name]! Thanks for your recent purchase!"
#     * Moderate: "Hi [name], we miss you! Check out our new products."
#     * At Risk: "Hi [name], it's been a while! Here's a special offer for you."
# Hint: Use str_extract() with pattern "^\\\\w+" for first name
# Hint: Use paste() to combine strings in case_when()

customer_outreach <- transactions_clean %>%
  mutate(
    # Your code here:
    
  )

# Display personalized messages
cat("Personalized Customer Messages:\n")
customer_outreach %>%
  select(customer_name, first_name, days_since, personalized_message) %>%
  head(10) %>%
  print()

In [ ]:
# Task 6.2: Analyze Transaction Patterns by Weekday
# TODO: Group by trans_weekday and calculate:
#   - transaction_count: number of transactions
#   - total_amount: sum of amount (if available)
#   - avg_amount: average amount per transaction
# TODO: Arrange by transaction_count descending

weekday_patterns <- transactions_clean %>%
  # Your code here:
  

# Display results
cat("Transaction Patterns by Weekday:\n")
print(weekday_patterns)

# Identify busiest day
busiest_day <- weekday_patterns$trans_weekday[1]
cat("\n🔥 Busiest day:", as.character(busiest_day), "\n")

In [ ]:
# Task 6.3: Monthly Transaction Analysis
# TODO: Group by trans_month_name and calculate:
#   - transaction_count
#   - unique_customers: use n_distinct(customer_name)
# TODO: Arrange by trans_month (to show chronological order)

monthly_patterns <- transactions_clean %>%
  # Your code here:
  

# Display results
cat("Monthly Transaction Patterns:\n")
print(monthly_patterns)

## Part 7: Business Intelligence Summary

**Business Context:** Create an executive summary that combines all your analyses into actionable insights.

**Your Tasks:**
1. Calculate key metrics across all datasets
2. Identify top products and categories
3. Summarize customer sentiment
4. Provide data-driven recommendations

In [ ]:
# Task 7.1: Create Business Intelligence Dashboard

cat("\n", rep("=", 60), "\n")
cat("         BUSINESS INTELLIGENCE SUMMARY\n")
cat(rep("=", 60), "\n\n")

# Product Analysis
cat("📦 PRODUCT ANALYSIS\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total number of products
#   - Number of wireless products
#   - Number of premium products
#   - Most common category


# Customer Sentiment
cat("\n💬 CUSTOMER SENTIMENT\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total feedback entries
#   - Average sentiment score
#   - Percentage of positive reviews
#   - Percentage of negative reviews


# Transaction Patterns
cat("\n📊 TRANSACTION PATTERNS\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total transactions
#   - Date range (earliest to latest)
#   - Busiest weekday
#   - Weekend transaction percentage


# Customer Recency
cat("\n👥 CUSTOMER RECENCY\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Number of recent customers (< 30 days)
#   - Number of at-risk customers (> 90 days)
#   - Percentage needing re-engagement



In [ ]:
# Task 7.2: Identify Top Products by Category
# TODO: Group products by category_clean and count products in each
# TODO: Arrange by count descending
# TODO: Display top 5 categories

top_categories <- products_clean %>%
  # Your code here:
  

cat("Top Product Categories:\n")
print(top_categories)

## Part 8: Reflection Questions

Answer the following questions based on your analysis. Write your answers in the markdown cells below.

### Question 8.1: Data Quality Impact

**How did cleaning the text data (removing spaces, standardizing case) improve your ability to analyze the data? Provide specific examples from your homework.**

Your answer here:



### Question 8.2: Pattern Detection Value

**What business insights did you gain from detecting patterns in product names (wireless, premium, gaming)? How could a business use this information?**

Your answer here:



### Question 8.3: Date Analysis Importance

**Why is analyzing transaction dates by weekday and month important for business operations? Provide at least three specific business applications.**

Your answer here:

1. 
2. 
3. 


### Question 8.4: Customer Recency Strategy

**Based on your recency analysis, what specific actions would you recommend for customers in each category (Recent, Moderate, At Risk)? How would you prioritize these actions?**

Your answer here:



### Question 8.5: Sentiment Analysis Application

**How could the sentiment analysis you performed be used to improve products or customer service? What are the limitations of this simple sentiment analysis approach?**

Your answer here:



### Question 8.6: Real-World Application

**Describe a real business scenario where you would need to combine string manipulation and date analysis (like you did in this homework). What insights would you be trying to discover?**

Your answer here:



## Summary and Submission

### What You've Accomplished

In this homework, you've successfully:
- ✅ Cleaned and standardized messy text data using `stringr` functions
- ✅ Detected patterns and extracted information from text
- ✅ Parsed dates and extracted temporal components using `lubridate`
- ✅ Calculated customer recency for segmentation
- ✅ Analyzed transaction patterns by time periods
- ✅ Combined string and date operations for business insights
- ✅ Created personalized customer communications
- ✅ Generated executive-ready business intelligence summaries

### Key Skills Mastered

**String Manipulation:**
- `str_trim()`, `str_squish()` - Whitespace handling
- `str_to_lower()`, `str_to_upper()`, `str_to_title()` - Case conversion
- `str_detect()` - Pattern detection
- `str_extract()` - Information extraction
- `str_count()` - Pattern counting

**Date/Time Operations:**
- `ymd()`, `mdy()`, `dmy()` - Date parsing
- `year()`, `month()`, `day()`, `wday()` - Component extraction
- `quarter()` - Period extraction
- `today()` - Current date
- Date arithmetic - Calculating differences

**Business Applications:**
- Data cleaning and standardization
- Customer segmentation by recency
- Sentiment analysis
- Pattern identification
- Temporal trend analysis
- Personalized communication

### Submission Checklist

Before submitting, ensure you have:
- [ ] Entered your name, student ID, and date at the top
- [ ] Completed all code tasks (Parts 1-7)
- [ ] Run all cells successfully without errors
- [ ] Answered all reflection questions (Part 8)
- [ ] Used proper commenting in your code
- [ ] Used the pipe operator (`%>%`) where appropriate
- [ ] Verified your results make business sense
- [ ] Checked for any remaining TODO comments

### Grading Criteria

Your homework will be evaluated on:
- **Code Correctness (40%)**: All tasks completed correctly
- **Code Quality (20%)**: Clean, well-commented, efficient code
- **Business Understanding (20%)**: Demonstrates understanding of business applications
- **Reflection Questions (15%)**: Thoughtful, complete answers
- **Presentation (5%)**: Professional formatting and organization

### Next Steps

In Lesson 8, you'll learn:
- Advanced data wrangling with complex pipelines
- Sophisticated conditional logic with `case_when()`
- Data validation and quality checks
- Creating reproducible analysis workflows
- Professional best practices for business analytics

**Great work on completing this assignment! 🎉**